File: clojure/proverb/src/proverb.clj
(ns proverb)

(defn recite [items]
  (if (empty? items) 
    ""
    (clojure.string/join "\n" (conj
     (mapv #(str "For want of a " (first %) " the " (last %) " was lost.") (partition 2 1 items))
     (str "And all for the want of a " (first items) ".")))))

File: elixir/proverb/lib/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """
  @spec recite(strings :: [String.t()]) :: String.t()
  def recite([]), do: ""
  def recite([first | _rest] = strings), do:
    strings
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.map(fn [want, lost] -> "For want of a #{want} the #{lost} was lost.\n" end)
    |> Enum.join()
    |> Kernel.<>("And all for the want of a #{first}.\n")
end
